### ISSUE TITLE
[Bug]: "Reveal" action highlights file in Finder but does not bring Finder to the foreground

### ISSUE THREAD

**Issue by lgarron** [quicksilver/Quicksilver#3017]
### Before submitting your bug report, please confirm you have completed the following steps

- [X] Searched the issue is not already listed in the [Issue List](https://github.com/quicksilver/Quicksilver/issues?q=is%3Aissue) (I've searched it)
- [X] Searched in the [Quicksilver Users' Group](https://groups.google.com/forum/#!forum/blacktree-quicksilver) for questions relating to tech support.
- [X] Downloaded the [latest version of Quicksilver](https://qsapp.com/) for your system and confirmed the issue still persists.

### Bug description

The "Reveal" action highlights file in Finder but does not bring Finder to the foreground

### Steps to reproduce

1. Navigate to a file in QS
2. Execute the "Reveal" action

### Expected behavior

The file is revealed and Finder is brought to the foreground.

### MacOS Version

Other

### Quicksilver Version

2.4.2

**Comment by lgarron** [comment]
I actually resolved this while filing the issue, but I thought I'd file it anyhow. That way, if anyone (including me 😅) runs into this in the future, hopefully they'll find this!

What worked for me was to go to System Settings → Privacy & Security and:

- Delete Quicksilver from "Full Disk Access" and add it again.
- Toggle the following permissions from Quicksilver off and on again:
  - Accessibility
  - Automation
  - Screen & System Audio Recording

I think this what caused it to work again, but I'm not sure which setting was borked.

**Comment by lgarron** [comment]
Hmm, I can't figure it out. Toggling the same permissions didn't work this time.

I've also observed another weird thing: The "Reveal" action *does* bring Finder ahead of *almost* all apps… except the current app. It's as if the current app either stays in the foreground, or something is bringing it back to the foreground before any effect becomes visible on the screen. 🤔

**Comment by n8henrie** [comment]
`Reveal` seems to be working as expected for me.

I currently have Firefox open in full screen and Books.app maximized but not full screen -- it moves focus away from Firefox and opens a Finder window in the next available space, and it pops up a Finder window in front of Books.

Is this a new behavior?

What is your MacOS version? (Sorry the template is outdated.)

**Comment by lgarron** [comment]
> Is this a new behavior?

Yeah, this has been happening for the last few weeks, quite possibly at the time of the Sequoia upgrade.

I don't think it was a consistent issue, as evidenced by the fact that I was temporarily able to fix it. But it seems to be fairly persistent now. :-/

> What is your MacOS version? (Sorry the template is outdated.)

15.0.1 (Sequoia)

**Comment by lgarron** [comment]
> What worked for me was to go to System Settings → Privacy & Security and:
> 
> * Delete Quicksilver from "Full Disk Access" and add it again.
> * Toggle the following permissions from Quicksilver off and on again:
>   
>   * Accessibility
>   * Automation
>   * Screen & System Audio Recording

Okay, going through these settings worked again this time. Let's see how long it sticks.

**Comment by lgarron** [comment]
> Okay, going through these settings worked again this time. Let's see how long it sticks.

Seems to have already regressed after 2 hours. 😕

**Comment by lgarron** [comment]
I just got the latest Quicksilver update, and [at least one of my actions](https://github.com/lgarron/dotfiles/blob/main/dotfiles/quicksilver/Library/Application%20Support/Quicksilver/Actions/Open%20folder%20in%20terminal%20app.scpt) broke as a result.

I tried the following:

> * Delete Quicksilver from "Full Disk Access" and add it again.
> * Toggle the following permissions from Quicksilver off and on again:
>   
>   * Accessibility
>   * Automation
>   * Screen & System Audio Recording

… but it's still broken. 😕

I'm sure it's not Quicksilver's "fault", but it seems the new binary lost the permissions of the old in a way that doesn't properly reflect in the system settings.

**Comment by lgarron** [comment]
> I just got the latest Quicksilver update, and [at least one of my actions](https://github.com/lgarron/dotfiles/blob/main/dotfiles/quicksilver/Library/Application%20Support/Quicksilver/Actions/Open%20folder%20in%20terminal%20app.scpt) broke as a result.
> 
> I tried the following:
> 
> > * Delete Quicksilver from "Full Disk Access" and add it again.
> > * Toggle the following permissions from Quicksilver off and on again:
> >   
> >   * Accessibility
> >   * Automation
> >   * Screen & System Audio Recording
> 
> … but it's still broken. 😕
> 
> I'm sure it's not Quicksilver's "fault", but it seems the new binary lost the permissions of the old in a way that doesn't properly reflect in the system settings.

Okay, I tried `sudo reboot now` to fix this and that seems to have done the job. I can't make sense of macOS permissions anymore. 😕

**Comment by pjrobertson** [comment]
If you have more than one copy of Quicksilver on your disk, it can mess with macOS permissions. The most reliable way I've found is to delete and re-add Quicksilver (the right one!) to the various permissions lists.

Is this original issue still valid, or can it be closed?

**Comment by lgarron** [comment]
> If you have more than one copy of Quicksilver on your disk, it can mess with macOS permissions. 

Hmm, does this apply to automatic Quicksilver updates? Does that leave an old copy on disk somewhere?

I certainly haven't added or removed any Quicksilver binaries on the system myself.

> The most reliable way I've found is to delete and re-add Quicksilver (the right one!) to the various permissions lists.

Yeah, as noted above that simply doesn't work for me anymore. I tried removing and adding Quicksilver to every permission list and restarting QS, and that did nothing until I had a chance to reboot the whole system. 😕

> Is this original issue still valid, or can it be closed?

Yeah, it's still an issue. 😭

It also seems to be 1) flaky, 2) not limited to Quicksilver, 3) happening on multiple devices. That makes me think it's one of those bugs where we're at the whim of Apple. It's gotten to the point that I've added workarounds to some of my programs to manually bring Finder to the foreground, but I still have no idea what causes it.

Is there any way to bring a workaround like that to Quicksilver, without messing things up for those who don't have the issue?

**Comment by n8henrie** [comment]
A few quick thoughts
- toggling permissions def don't work. Must be remove (with the minus sign) an re-add
- tccutil and looking at the respective database are sometimes interesting for these issues
- should not be an issue for QS updates in general, **unless we change / update our signing cert**
- macOS **does not like** having the same bundle ID signed by different certs when it comes to special permissions. Relevant for anyone running any combination of running source / developer builds, debug builds, ad-hoc certs, and final / production builds

I have seldom if ever needed to reboot to resolve these issues, but I've also seen enough quirky adjacent behavior that I don't doubt the story.

In general, when things get funky, the process should be to
1. Quit QS
2. Ensure all QS processes have ended (activity monitor)
4. **remove** QS from the permissions in question (with the `-` in the lower left corner, toggling the button won't cut it)
5. restart QS
6. re-accept permissions

Thankfully this should be really rare -- as mentioned, macOS seems to be happy as long as the bundle ID is identified with the same cert.

**Comment by lgarron** [comment]
I just updated to the latest Quicksilver (2.6.0) and it seems Quicksilver lost permissions again at upgrade time. (macOS 26.4.1)

In this case, I was able to fix it using just:

1. System Settings → Privacy & Security → Accessibility → Quicksilver → Toggle on and off
2. Relaunch Quicksilver
